File: packages/calendar/src/__tests__/CalendarModule.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { RynBridge, MockTransport } from '@rynbridge/core';
import type { BridgeResponse } from '@rynbridge/core';
import { CalendarModule } from '../CalendarModule.js';

describe('CalendarModule', () => {
  let transport: MockTransport;
  let bridge: RynBridge;
  let calendar: CalendarModule;

  beforeEach(() => {
    transport = new MockTransport();
    bridge = new RynBridge({ timeout: 5000 }, transport);
    calendar = new CalendarModule(bridge);
  });

  function respondSuccess(payload: Record<string, unknown> = {}) {
    const sent = JSON.parse(transport.sent[transport.sent.length - 1]);
    const response: BridgeResponse = {
      id: sent.id,
      status: 'success',
      payload,
      error: null,
    };
    transport.simulateIncoming(JSON.stringify(response));
  }

  describe('getCalendars', () => {
    it('returns list of calendars', async () => {
      const promise = calendar.getCalendars();
      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('calendar');
      expect(sent.action).toBe('getCalendars');

      respondSuccess({ calendars: [{ id: 'cal-1', title: 'Work', color: '#FF0000', isReadOnly: false }] });
      const result = await promise;
      expect(result.calendars).toHaveLength(1);
      expect(result.calendars[0].title).toBe('Work');
    });
  });

  describe('getEvents', () => {
    it('returns events in date range', async () => {
      const promise = calendar.getEvents({ from: '2026-01-01T00:00:00Z', to: '2026-01-31T23:59:59Z' });
      respondSuccess({ events: [{ id: 'evt-1', calendarId: 'cal-1', title: 'Meeting', startDate: '2026-01-15T10:00:00Z', endDate: '2026-01-15T11:00:00Z', isAllDay: false }] });
      const result = await promise;
      expect(result.events).toHaveLength(1);
      expect(result.events[0].title).toBe('Meeting');
    });
  });

  describe('createEvent', () => {
    it('creates an event and returns id', async () => {
      const promise = calendar.createEvent({
        title: 'New Event',
        startDate: '2026-03-01T10:00:00Z',
        endDate: '2026-03-01T11:00:00Z',
      });
      respondSuccess({ id: 'evt-new' });
      const result = await promise;
      expect(result.id).toBe('evt-new');
    });
  });

  describe('deleteEvent', () => {
    it('deletes an event', async () => {
      const promise = calendar.deleteEvent({ id: 'evt-1' });
      respondSuccess();
      await promise;
    });
  });

  describe('createReminder', () => {
    it('creates a reminder and returns id', async () => {
      const promise = calendar.createReminder({ title: 'Remember this' });
      respondSuccess({ id: 'rem-1' });
      const result = await promise;
      expect(result.id).toBe('rem-1');
    });
  });

  describe('requestPermission', () => {
    it('requests calendar permission', async () => {
      const promise = calendar.requestPermission();
      respondSuccess({ granted: true });
      const result = await promise;
      expect(result.granted).toBe(true);
    });
  });

  describe('getPermissionStatus', () => {
    it('returns permission status', async () => {
      const promise = calendar.getPermissionStatus();
      respondSuccess({ status: 'granted' });
      const result = await promise;
      expect(result.status).toBe('granted');
    });
  });

  describe('error handling', () => {
    it('propagates bridge errors', async () => {
      const promise = calendar.getCalendars();
      const sent = JSON.parse(transport.sent[0]);
      const response: BridgeResponse = {
        id: sent.id,
        status: 'error',
        payload: {},
        error: { code: 'UNKNOWN', message: 'Calendar access denied' },
      };
      transport.simulateIncoming(JSON.stringify(response));
      await expect(promise).rejects.toThrow('Calendar access denied');
    });
  });
});

File: android/calendar/src/test/kotlin/com/devryner/rynbridge/calendar/CalendarModuleTest.kt
package com.devryner.rynbridge.calendar

import com.devryner.rynbridge.core.*
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

class CalendarModuleTest {

    @Test
    fun `getCalendars returns calendar list`() = runTest {
        val provider = MockCalendarProvider()
        val module = CalendarModule(provider)
        val handler = module.actions["getCalendars"]!!

        val result = handler(emptyMap())
        val calendars = result["calendars"]?.arrayValue
        assertNotNull(calendars)
        assertEquals(1, calendars!!.size)
        assertEquals("cal-1", calendars[0].dictionaryValue?.get("id")?.stringValue)
        assertEquals("Personal", calendars[0].dictionaryValue?.get("title")?.stringValue)
    }

    @Test
    fun `getEvents returns event list`() = runTest {
        val provider = MockCalendarProvider()
        val module = CalendarModule(provider)
        val handler = module.actions["getEvents"]!!

        val result = handler(mapOf(
            "calendarId" to BridgeValue.string("cal-1"),
            "from" to BridgeValue.string("2026-01-01"),
            "to" to BridgeValue.string("2026-12-31")
        ))
        val events = result["events"]?.arrayValue
        assertNotNull(events)
        assertEquals(1, events!!.size)
        assertEquals("Meeting", events[0].dictionaryValue?.get("title")?.stringValue)
    }

    @Test
    fun `getEvent returns single event`() = runTest {
        val provider = MockCalendarProvider()
        val module = CalendarModule(provider)
        val handler = module.actions["getEvent"]!!

        val result = handler(mapOf("id" to BridgeValue.string("evt-1")))
        assertEquals("evt-1", result["id"]?.stringValue)
        assertEquals("Meeting", result["title"]?.stringValue)
        assertEquals("2026-03-06T10:00:00", result["startDate"]?.stringValue)
    }

    @Test
    fun `createEvent returns id`() = runTest {
        val provider = MockCalendarProvider()
        val module = CalendarModule(provider)
        val handler = module.actions["createEvent"]!!

        val result = handler(mapOf(
            "calendarId" to BridgeValue.string("cal-1"),
            "title" to BridgeValue.string("New Event"),
            "startDate" to BridgeValue.string("2026-03-07T09:00:00"),
            "endDate" to BridgeValue.string("2026-03-07T10:00:00"),
            "location" to BridgeValue.string("Office"),
            "notes" to BridgeValue.string("Important"),
            "isAllDay" to BridgeValue.bool(false)
        ))
        assertEquals("new-evt", result["id"]?.stringValue)
    }

    @Test
    fun `updateEvent calls provider`() = runTest {
        val provider = MockCalendarProvider()
        val module = CalendarModule(provider)
        val handler = module.actions["updateEvent"]!!

        val result = handler(mapOf(
            "id" to BridgeValue.string("evt-1"),
            "title" to BridgeValue.string("Updated Meeting")
        ))
        assertTrue(result.isEmpty())
        assertEquals("evt-1", provider.lastUpdateId)
        assertEquals("Updated Meeting", provider.lastUpdateTitle)
    }

    @Test
    fun `deleteEvent calls provider`() = runTest {
        val provider = MockCalendarProvider()
        val module = CalendarModule(provider)
        val handler = module.actions["deleteEvent"]!!

        val result = handler(mapOf("id" to BridgeValue.string("evt-1")))
        assertTrue(result.isEmpty())
        assertEquals("evt-1", provider.lastDeleteId)
    }

    @Test
    fun `createReminder returns id`() = runTest {
        val provider = MockCalendarProvider()
        val module = CalendarModule(provider)
        val handler = module.actions["createReminder"]!!

        val result = handler(mapOf(
            "title" to BridgeValue.string("Buy groceries"),
            "dueDate" to BridgeValue.string("2026-03-07"),
            "notes" to BridgeValue.string("Milk, eggs")
        ))
        assertEquals("rem-1", result["id"]?.stringValue)
    }

    @Test
    fun `requestPermission returns granted`() = runTest {
        val provider = MockCalendarProvider()
        val module = CalendarModule(provider)
        val handler = module.actions["requestPermission"]!!

        val result = handler(emptyMap())
        assertEquals(true, result["granted"]?.boolValue)
    }

    @Test
    fun `getPermissionStatus returns status`() = runTest {
        val provider = MockCalendarProvider()
        val module = CalendarModule(provider)
        val handler = module.actions["getPermissionStatus"]!!

        val result = handler(emptyMap())
        assertEquals("granted", result["status"]?.stringValue)
    }

    @Test
    fun `module name and version`() {
        val provider = MockCalendarProvider()
        val module = CalendarModule(provider)
        assertEquals("calendar", module.name)
        assertEquals("0.1.0", module.version)
    }
}

private class MockCalendarProvider : CalendarProvider {
    var lastUpdateId: String? = null
    var lastUpdateTitle: String? = null
    var lastDeleteId: String? = null

    private val mockCalendar = CalendarData(
        id = "cal-1",
        title = "Personal",
        color = "#FF0000",
        isReadOnly = false
    )

    private val mockEvent = CalendarEventData(
        id = "evt-1",
        calendarId = "cal-1",
        title = "Meeting",
        startDate = "2026-03-06T10:00:00",
        endDate = "2026-03-06T11:00:00",
        location = "Room A",
        notes = "Discuss project",
        isAllDay = false
    )

    override suspend fun getCalendars(): List<CalendarData> = listOf(mockCalendar)

    override suspend fun getEvents(calendarId: String?, from: String, to: String): List<CalendarEventData> =
        listOf(mockEvent)

    override suspend fun getEvent(id: String): CalendarEventData = mockEvent

    override suspend fun createEvent(
        calendarId: String?,
        title: String,
        startDate: String,
        endDate: String,
        location: String?,
        notes: String?,
        isAllDay: Boolean
    ): String = "new-evt"

    override suspend fun updateEvent(
        id: String,
        title: String?,
        startDate: String?,
        endDate: String?,
        location: String?,
        notes: String?,
        isAllDay: Boolean?
    ) {
        lastUpdateId = id
        lastUpdateTitle = title
    }

    override suspend fun deleteEvent(id: String) {
        lastDeleteId = id
    }

    override suspend fun createReminder(title: String, dueDate: String?, notes: String?): String = "rem-1"

    override suspend fun requestPermission(): Boolean = true

    override suspend fun getPermissionStatus(): String = "granted"
}
